### DISCUSSION TITLE
Speed comparison: Github Actions takes 3m 32s, Circle CI takes 12s

### DISCUSSION THREAD

**Discussion by g105b** [community/community#26146] [Actions]
<p>Hi all,</p>
<p>I’m looking for some advice of how to speed Github Actions up. I have always used Circle CI to perform my CI tests as I like how it utilises layering of containers to achieve good speeds. A fast build is something I value highly.</p>
<p>Please see this build run on Circle CI, where I merged branch ‘master’ into branch ‘github-actions’: <a href="https://circleci.com/gh/PhpGt/Dom/528" rel="nofollow noopener">https://circleci.com/gh/PhpGt/Dom/528</a></p>
<p>The whole build and test run took only 12 seconds to complete. It even uploaded all of the test coverage artifacts so you can <a href="https://528-49077458-gh.circle-artifacts.com/0/TestCoverage/index.html" rel="nofollow noopener">view them online</a>.</p>
<p>Compare the same test execution when run on Github actions: <a href="https://github.com/PhpGt/Dom/actions/runs/24972497" rel="nofollow noopener">https://github.com/PhpGt/Dom/actions/runs/24972497</a>.</p>
<p>Github Actions takes 3 minutes and 32 seconds to compelte the same task. I know I’m comparing apples to oranges here, but the difference in execution time is so significant, I’m sure I am doing something wrong.</p>
<p>Something to note is that I break my build and test steps into different jobs. This is so that when projects become more complex and have different test runners, they can run concurrently – once the dependencies are built, PHPUnit tests can run simultaneously with the Behat and Codeception, for example (the linked project uses only PHPUnit for now).</p>
<p>Looking into the specifics of what makes Circle CI so fast, I have found the following points:</p>
<ol>
<li>Circle CI cache their built layers. If the infrastructure doesn’t change, such as version of PHP, version of Linux, etc. then the same layers can be used across multiple builds. It takes 2 seconds to “Spin up the environment”.</li>
<li>It’s possible to cache the dependencies so that if the dependencies do not change between builds, the pre-built “vendor/” directory will be cached. For non-PHP people, this is identical to the concept of the “node_modules/” directory, where I expect this issue is even more prevalent.</li>
<li>You can see the caching set up here: <a href="https://github.com/PhpGt/Dom/blob/github-actions/.circleci/config.yml#L13-L23" rel="nofollow noopener">https://github.com/PhpGt/Dom/blob/github-actions/.circleci/config.yml#L13-L23</a>
</li>
<li>I have tried to do the same in Github Actions here: <a href="https://github.com/PhpGt/Dom/blob/github-actions/.github/workflows/ci.yml#L12-L16" rel="nofollow noopener">https://github.com/PhpGt/Dom/blob/github-actions/.github/workflows/ci.yml#L12-L16</a>
</li>
<li>The saving and restoring of the</li>
</ol>
<p>Thanks everyone!</p>
<p>Greg.</p>

**Comment by Yanjingzhu** [comment]
<p>In Github Actions, when you choose Github Hosted Runner to run your workflows, it provides a clean instance for every job execution without built layers nor dependencies .</p>
<p>If the <a href="https://github.com/actions/cache" rel="nofollow noopener">actions/cache</a> could not reach your requirement, you could consider to use <a href="https://help.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners" rel="nofollow noopener">self-hosted runner</a>, then the built layers and dependencies will be stored in your machine which installed Github Action Runner App . </p>

**Comment by g105b** [comment]
<p><span class="mention">@yanjingzhuthank</span> you for your reply. Github strongly advise <strong>not to use</strong> self-hosted runners for open source development (<a href="https://help.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#self-hosted-runner-security-with-public-repositories" rel="nofollow noopener">https://help.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners#self-hosted-runner-security-with-public-repositories</a>).</p>
<p>If what you say is correct, then Github Actions are not suitable for open source development. That is a shame.</p>

**Comment by rlittlefield** [comment]
<p>Edit: Just realized you are the one who made that example of caching the dependencies. See comment below.</p>
<p>Definitely look into the actions cache. You can use the actions/cache for composer’s vendor directory, using a hash of the composer.lock file as the cache key. Here is an example of a github action that shows using the cache this same way.</p>
<pre><code>-name:Cache PHP dependenciesuses:actions/cache@v1with:
path:vendorkey:${{ runner.OS }}-build-${{ hashFiles('\*\*/composer.lock') }}
</code></pre>
<p>(from <a href="https://github.com/marketplace/actions/composer-php-actions" rel="nofollow noopener">https://github.com/marketplace/actions/composer-php-actions</a>)</p>
<p>I don’t know if you will be able to get it down to 12 seconds like with Circle, but if most of your time is spent on dependencies it could help a lot.</p>

**Comment by rlittlefield** [comment]
<p>I just saw that you are already using the cache (and are apparently the source of where the cache example is coming from). My bad. The logs from your build are gone now so I couldn’t see the breakdown of where all the time was spent.</p>

**Comment by Yanjingzhu** [comment]
<p>Oh, sorry for missing that you are using a public repo. Yes, github actions doesn’t suggest to use self-hosted runner for public repo. Using self-hosted runner is not a good suggestion. </p>
<p>Which step cost the most of time to run in github action? I checked your repo, in the latest build “Use Github hosted runner to isolate issue” it is the upload artifacts (all your source code) and download artifacts that cost 2 more minutes. And I checked your circle ci build again , you just upload test results and coverage report. </p>

**Comment by DannyBen** [comment, marked as answer]
<p>I believe GitHub will have to find a solution to docker layer caching. This seems to be the number one reason for “my build is slower on GitHub Actions” complaints. We are suffering from this greatly, but the enjoyment of using GitHub Actions eases the pain a little.</p>
<p>I believe docker layer caching is discussed in several places, not sure which of them will get traction and attention. Perhaps join the discussion / upvote one of these places / send feedback to GitHub?</p>
<p>Here are two:</p>
<p>- <a href="https://github.com/actions/cache/issues/31" rel="nofollow noopener">actions/cache #31 - Docker caching</a></p>
<ul>
<li><a href="https://github.community/t/docker-layer-caching/17212">Docker layer cachine (community forum)</a></li>
</ul>

**Comment by g105b** [comment]
<p><span class="mention">@rlittlefieldI</span> already use actions/cache to cache Composer’s vendor directory. My question is in relation to persisting files between jobs, which actions/cache does not do, as far as I can tell.</p>

**Comment by g105b** [comment]
<p><span class="mention">@yanjingzhuPlease</span> check my Circle CI build again. I have merged the Circle CI config into the same branch, github-actions. Every single file is persisted between “build” job and “test-phpunit” job. Both jobs execute and complete within 30 seconds combined.</p>
<p>Also, uploading “all my source code” shouldn’t take 2 minutes - where is it uploading to? Surely it’s the same box? All of my source code consists of no more than a few megabytes of text files.</p>

**Comment by g105b** [comment]
<p><span class="mention">@dannybenyou</span> are right. Without layer caching, the product will always be slow. There is nothing we can do about it. I will contribute to the links you provided. Thanks very much for your response.</p>

**Comment by Yanjingzhu** [comment]
<p>Upload artifacts will upload to github server side.  I tried to fork your repo , I found that it is <a href="https://github.com/php-actions/composer" rel="nofollow noopener">php-actions/composer</a> action slows down the upload artifacts step, but I could not figure out why this happens.  </p>

**Comment by rlittlefield** [comment]
<p>I was under the impression that actions/cache could persist files between jobs, as long as you have the same cache key. Isn’t that the main idea?</p>
<p>As far as docker layer caching, if you are doing it for your own docker builds, I’ve found an action that will run the builds for you, and will use a connection to a container registry to dump all the layers. I’m using it with google’s <a href="http://gcr.io" rel="nofollow noopener">gcr.io</a>, and it works fine:</p>
<p><a href="https://github.com/whoan/docker-build-with-cache-action" rel="nofollow noopener">https://github.com/whoan/docker-build-with-cache-action</a></p>
